lxd/db/openfga.go

```diff
@@ -0,0 +1,610 @@
+//
+// Notes:
+//   - The exception for the type-bound public access may be better placed as a contextual tuple. However, adding it here
+//     means we can avoid an unnecessary transaction that will happen a lot.
+//   - We will need to modify this exception when we implement service accounts.
+func (o *openfgaStore) ReadUsersetTuples(ctx context.Context, store string, filter storage.ReadUsersetTuplesFilter) (storage.TupleIterator, error) {
+	// Expect both an object and a relation.
+	if filter.Object == "" || filter.Relation == "" {
+		return nil, fmt.Errorf("ReadUsersetTuples: Filter must include both an object and a relation")
+	}
+
+	// Expect a URL to be present for the object. (E.g. we don't want to list all groups that have `can_view` on
+	// all projects, we should be checking for a specific entity).
+	entityTypeStr, entityURL, hasURL := strings.Cut(filter.Object, ":")
+	if !hasURL {
+		return nil, fmt.Errorf("ReadUsersetTuples: Listing all entities of type not supported")
+	}
+
+	entityType := entity.Type(entityTypeStr)
+	err := entityType.Validate()
+	if err != nil {
+		return nil, fmt.Errorf("ReadUsersetTuples: Invalid object filter %q: %w", filter.Object, err)
+	}
+
+	// Check for type-bound public access exception.
+	if entityType == entity.TypeServer && filter.Relation == "can_view" {
+		return storage.NewStaticTupleIterator([]*openfgav1.Tuple{
+			// Only returning one tuple here for the identity. When adding service accounts, we'll need
+			// to add another tuple to account for them.
+			{
+				Key: &openfgav1.TupleKey{
+					Object:   fmt.Sprintf("%s:%s", entity.TypeServer, entity.ServerURL().String()),
+					Relation: string(auth.EntitlementCanView),
+					User:     fmt.Sprintf("%s:*", entity.TypeIdentity),
+				},
+			},
+		}), nil
+	}
+
+	u, err := url.Parse(entityURL)
+	if err != nil {
+		return nil, fmt.Errorf("ReadUsersetTuples: Failed to parse entity URL %q: %w", entityURL, err)
+	}
+
+	var groupNames []string
+	err = o.clusterDB.Transaction(ctx, func(ctx context.Context, tx *ClusterTx) error {
+		// Get the ID of the entity.
+		entityRef, err := cluster.GetEntityReferenceFromURL(ctx, tx.Tx(), &api.URL{URL: *u})
+		if err != nil {
+			return err
+		}
+
+		// Get all groups with the permission.
+		q := `
+SELECT auth_groups.name 
+FROM auth_groups_permissions
+JOIN auth_groups ON auth_groups_permissions.auth_group_id = auth_groups.id
+WHERE auth_groups_permissions.entitlement = ? AND auth_groups_permissions.entity_type = ? AND auth_groups_permissions.entity_id = ?
+`
+		groupNames, err = query.SelectStrings(ctx, tx.Tx(), q, filter.Relation, cluster.EntityType(entityType), entityRef.EntityID)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+	if err != nil && !api.StatusErrorCheck(err, http.StatusNotFound) {
+		return nil, err
+	} else if err != nil {
+		// If we have a not found error then there are no tuples to return, but the datastore shouldn't return an error.
```

Comment:
This is quite awkwardly structured, where the specific scenario that occurs when there is a not found error is checking err != nil. In my view clearer would have been:

```go
if err != nil {
        if api.StatusErrorCheck(err, http.StatusNotFound) {
	        // If we have a not found error then there are no tuples to return, but the datastore shouldn't return an error.
		    return storage.NewStaticTupleIterator(nil), nil
        }

		return nil, err
	} 
```